test/merkle.js

```diff
@@ -0,0 +1,227 @@
+import { deployContract } from './utils';
+
+const OP = require('./helpers/constants');
+const ethers = require('ethers');
+
+const EthereumRuntime = artifacts.require('EthereumRuntime.sol');
+
+function getOpcodeName(op) {
+  for (var key in OP) {
+    if (OP[key] === op) {
+      return key;
+    }
+  }
+
+  return 'unknown opcode: ' + op;
+}
+
+function compactStack(before, after) {
+  let stackNeeded = [];
+  let len = before.length;
+
+  if (after.length > len) {
+    len = after.length;
+  }
+
+  for (var i = 0; i < len; i++) {
```

Comment:
I think you can for from len to 0
as most of the case 3 elements is needed
at most 16 elements are needed

Reply:
@peara Ya, had to count from 0 to find the first difference, after that I can break out.
But I need to add special checks for SWAPx and DUPx opcodes before the loop.

If we for example SWAP1 (0, 0) the stack looks the same before and after, but it did change anyway; so I need special handling for those cases 😄

Reply:
@peara Taking a look again. That is still a better idea you proposed 👍 .
Sometimes I'm reading the text wrong 😉 🍷

Reply:
@pinkiebell I also looked again! I guess the best way is just provide a const value for the number of required elements in stack depends on opcode. We may even enforce it at the runtime contract.

Reply:
@peara Yes! We can also declare the number of stack pops & stack pushes for each opcode.